src/main/java/org/spongepowered/common/world/pregen/SpongeChunkPreGenerateTask.java

```diff
@@ -0,0 +1,470 @@
+        final long stepStartTime = System.currentTimeMillis();
+        if (this.generationStartTime == 0) {
+            this.generationStartTime = stepStartTime;
+        }
+
+        // Create and fire event.
+        ChunkPreGenerationEvent.Pre preEvent = SpongeEventFactory.createChunkPreGenerationEventPre(
+                this.cause,
+                this,
+                this.world,
+                false
+        );
+
+        if (Sponge.getEventManager().post(preEvent)) {
+            // Cancelled event = cancelled task.
+            cancelTask(task);
+            return;
+        }
+
+        if (preEvent.getSkipStep()) {
+            // Skip the step, but don't cancel the task.
+            return;
+        }
+
+        // Count how many chunks are generated during the tick
+        int count = 0;
+        int skipped = 0;
+        do {
+            final Vector3i position = nextChunkPosition();
+            final Vector3i pos1 = position.sub(Vector3i.UNIT_X);
+            final Vector3i pos2 = position.sub(Vector3i.UNIT_Z);
+            final Vector3i pos3 = pos2.sub(Vector3i.UNIT_X);
+
+            // We can only skip generation if all chunks are loaded.
+            if (!areAllChunksLoaded(position, pos1, pos2, pos3)) {
+
+                // At least one chunk isn't generated, so to populate, we need to load them all.
+                this.world.loadChunk(position, true);
+                this.world.loadChunk(pos1, true);
+                this.world.loadChunk(pos2, true);
+                this.world.loadChunk(pos3, true);
+
+                count += this.currentGenCount;
+            } else {
+
+                // Skipped them, log this.
+                skipped += this.currentGenCount;
+            }
+        } while (hasNextChunkPosition() && checkChunkCount(count) && checkTickTime(System.currentTimeMillis() - stepStartTime));
+
+        this.chunksGenerated += count;
+        this.chunksSkipped += skipped;
+
+        final long deltaTime = System.currentTimeMillis() - stepStartTime;
+        this.generationEndTime = System.currentTimeMillis();
+
+        // Create and fire event.
+        if (Sponge.getEventManager().post(SpongeEventFactory.createChunkPreGenerationEventPost(
+                this.cause,
+                this,
+                this.world,
+                Duration.ofMillis(deltaTime),
+                count,
+                skipped
+        ))) {
+            cancelTask(task);
+            return;
+        }
+
+        if (!hasNextChunkPosition()) {
+            // Generation has completed.
+            Sponge.getEventManager().post(SpongeEventFactory.createChunkPreGenerationEventComplete(
+                    this.cause,
+                    this,
+                    this.world
+            ));
+
+            this.isCancelled = true;
+            unregisterListener();
+            task.cancel();
+        }
+    }
+
+    private boolean areAllChunksLoaded(Vector3i... chunksToCheck) {
```

Comment:
Is this only used at line 247? Because if so, it should just take 4 chunks as an argument and bypass all the object creation from using varargs and streams. It complicates things for no good reason.